### PR TITLE
AllowOverflow option to avoid checking int range

### PR DESCRIPTION
**Pull request by Hecate2**
unchecked((int)bigInteger))
Will conflict. Draft for now.

### CODE REVIEW

**Comment by Jim8y** [comment]
This pr has problem

```C#
 byte x = (byte)-1;
```
Overflow in constant value computation.

**Comment by Hecate2** [comment]
Changed the methodology. Skip range checks and throws with compilation option

**Comment by Hecate2** [comment]
Need a way to compile neo-devpack-dotnet\tests\Neo.Compiler.CSharp.TestContracts\Contract_BigInteger.cs with SimulateOverflow=true

**Comment by Hecate2** [comment]
> Need a way to compile neo-devpack-dotnet\tests\Neo.Compiler.CSharp.TestContracts\Contract_BigInteger.cs with SimulateOverflow=true

Allowed CompilationOptions to be used in tests.
TestGasConsume is set to false.

File: src/Neo.Compiler.CSharp/Program.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ public static int Main(string[] args)
                 new Option<string>(new[] { "-o", "--output" }, "Specifies the output directory."),
                 new Option<string>("--base-name", "Specifies the base name of the output files."),
                 new Option<NullableContextOptions>("--nullable", () => NullableContextOptions.Annotations, "Represents the default state of nullable analysis in this compilation."),
+                new Option<bool>("--simulate-overflow", "Indicates whether to simulate overflow and underflow. If false, --checked does not take effect."),
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/Program.cs:44]: --allow-overflow ?

By default we should test the contracts without this option

**Comment by Jim8y** [comment]
This name is so scary....

**Comment by Jim8y** [comment]
this can save tons of gas, yes,,,,,, but,,,,,,,

**Comment by Jim8y** [comment]
rather than allow overflow, i think biginteger only is better

**Comment by shargon** [comment]
We should not save gas if the security is involved

**Comment by shargon** [comment]
Sorry miss click

**Comment by Hecate2** [comment]
> We should not save gas if the security is involved 

No security problem is involved. This is just handling all int, long, short as BigInteger.

**Comment by Hecate2** [comment]
> rather than allow overflow, i think biginteger only is better

Sometimes it's impossible to write BigInteger literally in codes. For example, shifting operations (>> and <<) in C# supports only int as the shifted count of bits. Then we have to write an explicit coversion (int) which wastes a range check.

**Comment by Hecate2** [comment]
And our current default settings (with or without my PR) are --checked=false (--allow-overflow=false). This allows silent overflows (without or without my PR). I think this is even more dangerous. Maybe it's better to have default as --checked=false --allow-overflow=true, or --checked=true --allow-overflow=false

**Comment by shargon** [comment]
we should detect `unchecked` keyword

**Comment by Hecate2** [comment]
> we should detect `unchecked` keyword

We are actually detecting it (with or without this PR). But by default everything is unchecked.

**Comment by Jim8y** [comment]
> > rather than allow overflow, i think biginteger only is better
> 
> Sometimes it's impossible to write BigInteger everywhere literally in codes. For example, shifting operations (>> and <<) in C# supports only int as the shifted count of bits. Also indexes of arrays have to be int. Then we have to write an explicit coversion (int) which wastes a range check.

excatly, just use what is allowed in biginteger.

**Comment by Jim8y** [comment]
@shargon please review

**Comment by Jim8y** [comment]
This is an option that should be false by default, so after you @shargon reviewed, please dont merge it, i will update the test to remove the option and revert gas change.

**Comment by Jim8y** [comment]
> I still don't know why we need this option, which is the use case?

where user can handle the integer range and no need to do automatic range check. currently, every numerous operation requires a range check, its expensive.

**Comment by shargon** [comment]
> > I still don't know why we need this option, which is the use case?
> 
> where user can handle the integer range and no need to do automatic range check. currently, every numerous operation requires a range check, its expensive.

expensive not because it's required for safety

File: tests/Neo.Compiler.CSharp.UnitTests/TestCleanup.cs [hunk 2]
```diff
@@ -24,13 +24,24 @@ public class TestCleanup : TestCleanupBase
         private static readonly string TestContractsPath = Path.GetFullPath(Path.Combine("..", "..", "..", "..", "Neo.Compiler.CSharp.TestContracts", "Neo.Compiler.CSharp.TestContracts.csproj"));
         private static readonly string RootPath = Path.GetPathRoot(TestContractsPath) ?? string.Empty;
 
-        private static readonly Lazy<CompilationEngine> _compilationEngine = new(() => new CompilationEngine(new CompilationOptions
+        public static readonly CompilationOptions DefaultTestCompilationOptions = new CompilationOptions
         {
+            AllowOverflow = true,
```
Comment (shargon) [inline tests/Neo.Compiler.CSharp.UnitTests/TestCleanup.cs:29]: We should leave the test as before, only create one test for this specific case

Comment (Hecate2) [reply tests/Neo.Compiler.CSharp.UnitTests/TestCleanup.cs]: Done

**Review by Jim8y — APPROVED** [review]

**Comment by shargon** [comment]
Can we discuss this pe in the meeting?

**Comment by Hecate2** [comment]
> Can we discuss this pe in the meeting?

That's really great

**Comment by Jim8y** [comment]
gas examples to demonstrate gas saving.

**Comment by shargon** [comment]
Take a look to https://github.com/neo-project/neo-devpack-dotnet/pull/1249

**Comment by Jim8y** [comment]
close as can be partially replaced by using BigInteger, though not fully.

**Pull request closed without merging**